### PR TITLE
Enable IDE0048: AddRequiredParentheses

### PR DESCRIPTION
**Pull request by xtqqczze**
https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide0047-ide0048

Follow-up to #13896, after `EnforceCodeStyleInBuild` was enabled in #13957.

### CODE REVIEW

File: src/System.Management.Automation/engine/hostifaces/LocalConnection.cs [hunk 1]
```diff
@@ -987,7 +987,7 @@ private void StopOrDisconnectAllJobs()
                 foreach (Job job in this.JobRepository.Jobs)
                 {
                     // Only stop or disconnect PowerShell jobs.
-                    if (job is PSRemotingJob == false)
+                    if ((job is PSRemotingJob) == false)
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/hostifaces/LocalConnection.cs:990]: I wonder we skipped the compare with false.

**Review by iSazonov — COMMENTED** [review]
The rule is contradictory: parentheses are good for complex expressions to eliminate errors, but no doubt they are superfluous for simple expressions where they worsen readability.

I think it is better to set "suggestion" for the rule.

File: src/System.Management.Automation/engine/hostifaces/LocalConnection.cs (hunk 1 above)
Comment (xtqqczze) [reply src/System.Management.Automation/engine/hostifaces/LocalConnection.cs:990]: #13334 will remove the redundant literal

**Comment by xtqqczze** [comment]
@iSazonov Could you leave more review comments to give examples of where readability has been worsened?

File: src/Microsoft.PowerShell.Commands.Management/commands/management/GetComputerInfoCommand.cs [hunk 2]
```diff
@@ -1159,13 +1159,13 @@ internal static DateTime UnixSecondsToDateTime(long seconds)
...
 #else
             const int DaysPerYear = 365;
-            const int DaysPer4Years = DaysPerYear * 4 + 1;
-            const int DaysPer100Years = DaysPer4Years * 25 - 1;
-            const int DaysPer400Years = DaysPer100Years * 4 + 1;
-            const int DaysTo1970 = DaysPer400Years * 4 + DaysPer100Years * 3 + DaysPer4Years * 17 + DaysPerYear;
+            const int DaysPer4Years = (DaysPerYear * 4) + 1;
+            const int DaysPer100Years = (DaysPer4Years * 25) - 1;
+            const int DaysPer400Years = (DaysPer100Years * 4) + 1;
+            const int DaysTo1970 = (DaysPer400Years * 4) + (DaysPer100Years * 3) + (DaysPer4Years * 17) + DaysPerYear;
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/GetComputerInfoCommand.cs:1165]: Less readability.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs [hunk 3]
```diff
@@ -805,7 +805,7 @@ private byte[] GetSendBuffer(int bufferSize)
 
             for (int i = 0; i < bufferSize; i++)
             {
-                sendBuffer[i] = (byte)((int)'a' + i % 23);
+                sendBuffer[i] = (byte)((int)'a' + (i % 23));
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs:808]: Less readability.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs [hunk 4]
```diff
@@ -1289,7 +1289,7 @@ private string BuildErrorMessage(Diagnostic diagnisticRecord)
                 var errorLineString = textLines[errorLineNumber].ToString();
                 var errorPosition = lineSpan.StartLinePosition.Character;
 
-                StringBuilder sb = new StringBuilder(diagnisticMessage.Length + errorLineString.Length * 2 + 4);
+                StringBuilder sb = new StringBuilder(diagnisticMessage.Length + (errorLineString.Length * 2) + 4);
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:1292]: Less readability.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetRandomCommand.cs [hunk 5]
```diff
@@ -385,7 +385,7 @@ private Int64 GetRandomInt64(Int64 minValue, Int64 maxValue)
                 randomUint64 &= mask;
             } while (uint64Diff <= randomUint64);
 
-            double randomNumber = minValue * 1.0 + randomUint64 * 1.0;
+            double randomNumber = (minValue * 1.0) + (randomUint64 * 1.0);
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetRandomCommand.cs:388]: Less readability.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/Set-PSBreakpoint.cs [hunk 6]
```diff
@@ -88,8 +88,8 @@ protected override void BeginProcessing()
             // whether the RemoteScript debug option is selected.
             if (this.Context.InternalHost.ExternalHost is System.Management.Automation.Remoting.ServerRemoteHost &&
                 ((this.Context.CurrentRunspace == null) || (this.Context.CurrentRunspace.Debugger == null) ||
-                 ((this.Context.CurrentRunspace.Debugger.DebugMode & DebugModes.RemoteScript) != DebugModes.RemoteScript) &&
-                  (this.Context.CurrentRunspace.Debugger.DebugMode != DebugModes.None)))
+                 (((this.Context.CurrentRunspace.Debugger.DebugMode & DebugModes.RemoteScript) != DebugModes.RemoteScript) &&
+                  (this.Context.CurrentRunspace.Debugger.DebugMode != DebugModes.None))))
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/Set-PSBreakpoint.cs:92]: Good. This protects from regressions.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs [hunk 7]
```diff
@@ -2006,8 +2006,8 @@ internal string ReadLineWithTabCompletion(Executor exec)
                     // the user hit tab up to the current cursor position after writing the completed text.
 
                     int deltaInput =
-                        (endOfInputCursorPos.Y * screenBufferSize.Width + endOfInputCursorPos.X)
-                        - (endOfCompletionCursorPos.Y * screenBufferSize.Width + endOfCompletionCursorPos.X);
+                        ((endOfInputCursorPos.Y * screenBufferSize.Width) + endOfInputCursorPos.X)
+                        - ((endOfCompletionCursorPos.Y * screenBufferSize.Width) + endOfCompletionCursorPos.X);
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs:2010]: Less readability.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs [hunk 8]
```diff
@@ -2066,7 +2066,7 @@ private void SendLeftArrows(int length)
                 up.Data.Keyboard.ExtraInfo = IntPtr.Zero;
 
                 inputs[2 * i] = down;
-                inputs[2 * i + 1] = up;
+                inputs[(2 * i) + 1] = up;
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs:2069]: I'd wonder if anybody wanted parentheses in such expressions.

File: src/System.Management.Automation/FormatAndOutput/common/OutputQueue.cs [hunk 9]
```diff
@@ -69,9 +69,9 @@ internal List<PacketInfoData> Add(PacketInfoData o)
             }
 
             // STATE TRANSITION: we are processing and we stop
-            if (_processingGroup &&
+            if ((_processingGroup &&
                 ((o is GroupEndData) ||
-                (_objectCount > 0) && (_currentObjectCount >= _objectCount)) ||
+                ((_objectCount > 0) && (_currentObjectCount >= _objectCount)))) ||
                 ((_groupingTimer != null) && (_groupingTimer.Elapsed > _groupingDuration))
```
Comment (iSazonov) [inline src/System.Management.Automation/FormatAndOutput/common/OutputQueue.cs:75]: Good to prevents regressions.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/GetComputerInfoCommand.cs (hunk 2 above)
Comment (xtqqczze) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/GetComputerInfoCommand.cs:1165]: For most people it is obvious that multiplication has a higher precedence than addition, so the parenthesis are probably not necessary. However while readability might not be any better, it is not really any worse.

Comment (xtqqczze) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/GetComputerInfoCommand.cs:1165]: Multiplicative > Additive

File: src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs (hunk 3 above)
Comment (xtqqczze) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs:808]: Multiplicative > Additive

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs (hunk 4 above)
Comment (xtqqczze) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:1292]: Multiplicative > Additive

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetRandomCommand.cs (hunk 5 above)
Comment (xtqqczze) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetRandomCommand.cs:388]: Multiplicative > Additive

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs (hunk 7 above)
Comment (xtqqczze) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs:2010]: Multiplicative > Additive

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs (hunk 8 above)
Comment (xtqqczze) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs:2069]: Multiplicative > Additive

**Comment by xtqqczze** [comment]
@iSazonov It seems all your examples of worsened readability are for the higher precedence of multiplicative vs additive operators.  

Unfortunately the analyzer does not provide configuration at this level, the closest option is `dotnet_style_parentheses_in_arithmetic_binary_operators` covering *, /, %, +, -, <<, >>, &, ^, |

**Comment by iSazonov** [comment]
This is a controversial rule that depends on context. For simple expressions, no additional parentheses are needed. For complex expressions, good formatting and descriptive names are more important. I suppose we can turn this off at all.

**Comment by xtqqczze** [comment]
> I suppose we can turn this off at all.

OK so we use a lesser severity, `suggestion` or `hidden`?

**Comment by iSazonov** [comment]
> OK so we use a lesser severity, `suggestion` or `hidden`?

Is there Quick fix in VS Code? If yes, I agree with "suggestion", otherwise it is better to hide.

**Comment by iSazonov** [comment]
@xtqqczze I close the PR since it is controversial. I suggest these rules to "suggestion" in new PR.

**Pull request closed without merging**